### PR TITLE
[Model][1/N] Support multiple poolers at model level

### PR DESCRIPTION
**Pull request by DarkLight1337**
## Essential Elements of an Effective PR Description Checklist
- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

FIX #21277

## Purpose

This PR enables models to support multiple poolers, by:

- Splitting up the task validation part of `Pooler.get_pooling_updates` into `Pooler.get_supported_tasks`.
- Introducing `DispatchPooler` which dispatches pooler calls to sub-poolers based on the pooling task.

Apart from enabling models to customize each pooling task, this would eventually also let us output intermediate results within the same request, e.g.:
  - Hidden states via the `encode` pooler with `ALL` pooling.
  - Logits by adapting `LogitsProcessor` into a pooler.

Future work:

- Multi-task support based on `DispatchPooler` is not enabled at the API level yet. We still check `model_config.supported_tasks` which is based on the value of `--task`. This will be addressed in the next PR.
- To avoid confusing with pooling task, we will rename `--task` to `--convert` and ask users to pass `--runner pooling` to use pooling models in the next PR, emitting a deprecation warning to users still using `--task`.
- Memory profiling is based on `list(Pooler.get_supported_tasks())[0]` instead of running profiling on every sub-pooler and getting the maximum memory usage. So, it is possible for OOM to occur during inference time. We can refine this in a later PR.
- Disabling chunked prefill is still based on the `pooling_type` in the pooler config instead of the actual poolers being used, because the poolers are only created after model construction. Can we disable chunked prefill inside the model runner after the model is initialized?
- For now, Sentence Transformers pooler config overrides the poolers for all tasks. We should only override the "main" pooler based on the model architecture.

Notes:

- The output of `LLM.encode` (Pooling API) for embedding and classification models (except for models with Sentence Transforming pooler config) is changed by this PR because the default `encode` pooler is different from the default `embed` and default `classification` pooler. While technically a breaking change, users aren't supposed to use this API for embedding and classification tasks in the first place anyway.

Other changes:

- Renamed `--task pooling` to `--task encode`. This was introduced very recently in #20771 and is not documented, so this change should barely affect any users.
- Split up `BertModel` into a pooling variant and a non pooling variant to keep the type checker happy.

cc @maxdebayser @noooop @22quinn 

## Test Plan

The existing tests should pass.

## (Optional) Documentation Update

Updated the Pooling Models page to explain the new mechanism.

### CODE REVIEW

File: vllm/model_executor/models/bert.py [hunk 1]
```diff
@@ -395,10 +396,43 @@ def load_weights(self, weights: Iterable[tuple[str,
...
         )
+
+        config = vllm_config.model_config.hf_config
+        self.pooler = BertPooler(config)
+
+    def load_weights(self, weights: Iterable[tuple[str,
+                                                   torch.Tensor]]) -> set[str]:
+        other_weights, loaded_stacked_params = self._load_weights(weights)
+
+        loader = AutoWeightsLoader(self, skip_prefixes=["pooler."])
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/model_executor/models/bert.py]: Nice catch!

File: vllm/model_executor/layers/pooler.py [hunk 2]
```diff
@@ -79,37 +84,81 @@ class Pooler(nn.Module, ABC):
...
+            return base_pooler
+
+        return ClassifierPooler(
+            pooling=base_pooler.pooling,
+            classifier=classifier,
+            act_fn=base_pooler.head.activation,
+        )
+
+    @abstractmethod
+    def get_supported_tasks(self) -> Set[PoolingTask]:
```
Comment (Isotr0py) [inline vllm/model_executor/layers/pooler.py:153]:
```suggestion
    def get_supported_tasks(self) -> set[PoolingTask]:
```

I remember that we can use `set` for typing check after python3.9?

**Review by Isotr0py — APPROVED** [review]
Overall LGTM! Just a nit.

File: vllm/model_executor/layers/pooler.py (hunk 2 above)
Comment (DarkLight1337) [reply vllm/model_executor/layers/pooler.py:153]: I intentionally use `collections.abc.Set` which is immutable unlike the normal `set`

**Comment by noooop** [comment]
> * Disabling chunked prefill is still based on the `pooling_type` in the pooler config instead of the actual poolers being used, because the poolers are only created after model construction. Can we disable chunked prefill inside the model runner after the model is initialized?

`disabling chunked prefill & auto prefix cache`  should  be controlled by attn_type rather than pooling_type. For instance, Alibaba-NLP/gte-Qwen2-1.5B-instruct uses encoder-only attention, which should have chunked prefill & auto prefix cache disabled.

**Review by maxdebayser — COMMENTED** [review]
LGTM
